README.md
# ticketing README

A microsoft visual studio code extension providing a lab help-request ticketing system. Created as part of a UCL Connected Learning Internship 2020.

## Features

Currently features a 'module admin' screen, allowing the addition, removal and editing of lab courses. Data is saved in a JSON object format. Check the module_admin_screen.png for an image. 

* Users may click on any cell in the table to make edits easily. Press 'enter' when finished to save your changes. 
* 'Add Module' button brings up a form to add module info, e.g. name, code, admins, num students ... etc. and saves 
* Table displaying module info dynamically reloads to display new/edited/removed content 

The important files are: 

* htmlStuff.js      -- which provides the HTML code for the webview
* constructors.js   -- javascript file with the methods to add, remove and edit modules & perform first time setup of JSON """database"""
* extension.js      -- javascript file which activates extension 
* modules.txt       -- file which stores the JSON module data

## Requirements

## Extension Settings

* command Start Ticket from cmd pallet activates the extension. 

## Known and Predicted Issues

* Table scroll issues (possibly) when adding lots of modules. 
* HTML file is rife with inline javascript. Will refactor all this to a separate file / constructors soon.

## Upcoming Tasks 
* Adding more useful functionality to the module sessions. Implementing a calendar to select the date and times of lab session and save these in the ""database"".  


-----------------------------------------------------------------------------------------------------------


CHANGELOG.md
# Change Log

All notable changes to the "ticketing" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

extension.js
const vscode = require('vscode'); 
const constructors = require('./constructors.js');

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed

/**
 * @param {vscode.ExtensionContext} context
 */
function activate(context) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	console.log('Congratulations, your extension "ticketing" is now active!');

	// The command has been defined in the package.json file
	// Now provide the implementation of the command with  registerCommand
	// The commandId parameter must match the command field in package.json
	// my attempt at adding a fresh new custom command to the extension! Seems to be working...
	let newCommand = vscode.commands.registerCommand('ticketing.start', function () {
		vscode.window.showInformationMessage("Activating ticketing system!");
		
		// change the random working directory location to be this extension 
		//process.chdir('c:/Users/Shadowbiscuit/Desktop/ticketing'); 
		process.chdir('./Desktop/ticketing'); 
		console.log("Current working dir: " + process.cwd())

		// create instance of webview welcome screen OBJECT
		var screen = new constructors.Welcome(context);
	
	});
	context.subscriptions.push(newCommand);
}
exports.activate = activate;

// this method is called when your extension is deactivated
function deactivate() {}

module.exports = {
	activate,
	deactivate
}


htmlStuff.mjs
function getWelcomeScreen () {
	return `<!DOCTYPE html>
	<html lang="en">
	<head>
		<script>
		const vscode = acquireVsCodeApi();
		</script>

		<meta charset="UTF-8">
		<meta name="viewport" content="width=device-width, initial-scale=1.0">
		
		<title>Module Manager</title>
		
		<style> 
		body {
			background-color: rgb(60,99,201);}

		h1 {
			position: absolute; 
			top: 30%;
			left: 50%;
			transform: translate(-50%, -50%);
			color: white;}
		
		

		button {
			position: absolute; 
			top: 45%;
			left: 50%;
			transform: translate(-50%, -50%);
			background-color: rgb(51, 255, 255);
			color: rgb(51, 54, 255);
			border: border-width border-style border-color|initial|inherit;
			border-color: rgb(51, 255, 255);
		}

		</style>
	</head>
	
	<body>
		<h1>Welcome to the Lab Ticketing System</h1>

		<button onclick="moduleAdd()">Module Admin</button>
		<script>function moduleAdd() {
			vscode.postMessage({command: "alert", text: "BUTTON PRESSED!"});
			}
		</script> 
	
	</body>
	</html>`
};

function getModScreen () {
	return `<!DOCTYPE html>
	<html lang="en">
	<head>
		<script>
		const vscode = acquireVsCodeApi();
		</script>

		<script>
			function submit () {
				var Module = {
				aModuleName     : document.getElementById('name').value,
				bModuleCode     : document.getElementById('code').value,
				cModuleAdmins   : document.getElementById('admins').value,
				dModuleTAs      : document.getElementById('TAs').value,
				eModuleStudents : document.getElementById('stud').value,
				fModuleReqType  : document.getElementById('reqType').value,
				gModuleGroups   : document.getElementById('groups').value,
				hModuleHrs      : document.getElementById('hours').value};


				vscode.postMessage({command: "newInfo", info: Module});

				//hide the module input screen, re-display add module button and current modules added  
				document.getElementById('form1').style.display = 'none';
				document.getElementById('button1').style.display = ''; 
				document.getElementById('showData').style.display ='';
			}
		</script>

		<script>
			function dispForm () {
				document.getElementById('form1').style.display = '';
				document.getElementById('button1').style.display = 'none'; 
				// hide the module table
				document.getElementById('showData').style.display = 'none'; 
			};
		</script>

		<script>
			// create event listener for table data fetch 
			window.addEventListener('message', event => {
				var message = event.data; 
				
				switch (message.command) {
					
					case 'drawTab':
					// call drawing function and pass the JSON data  
					console.log('Message recieved: ' +  JSON.stringify(message.info)); 
					createTab(message.info, message.draw); 
				}
			});
		</script>

		<meta charset="UTF-8">
		<meta name="viewport" content="width=device-width, initial-scale=1.0">
		
		<title>Module Manager</title>
		
		<style>
		.forms {
			/* Center the form on the page */
			
			width: 400px;
			position: absolute; 
			top: 400px;
			right: 740px;
			/* Form outline */
			padding: 0.5em;
			border: 1px solid #CCC;
			border-radius: 1em;
		}
		  
		ul {
			list-style: none;
			padding: 0;
			margin: 0;
		}
		  
		.forms li + li {
			margin-top: 1em;
		}
		  
		label {
			/* Uniform size & alignment */
			display: inline-block;
			width: 90px;
			text-align: right;
		} 
		body {
			background-color: rgb(60,99,201);}

		h1 {
			text-align: center;
			position: absolute; 
			top: 30%;
			left: 50%;
			transform: translate(-50%, -50%);
			color: white;}

		#noModMsg {
			text-align: center; 
			position: absolute; 
			top: 40%; 
			left: 50%;
			transform: translate(-50%, -50%);
		}

		#button1 {
			position: absolute; 
			top: 45%;
			left: 50%;
			transform: translate(-50%, -50%);
			background-color: rgb(51, 255, 255);
			color: rgb(51, 54, 255);
			border: border-width border-style border-color|initial|inherit;
			border-color: rgb(51, 255, 255);
		}
		#button2 {
			position: relative;
			left: 52%;
		}

		table {
			position: absolute; 
			top: 50%; 
			left: 31%;
			border-collapse: collapse;}
		tr { 
			border: solid;
			border-width: 1px 0;
			}
		th, td{ 
			border-left:1px solid #ccc; 
			border-right: 1px solid #ccc;
		}

		.rmBtn {
			background-color: rgb(51, 255, 255);
			color: rgb(51, 54, 255);
			border: border-width border-style border-color|initial|inherit;
			border-color: rgb(51, 255, 255);
		}
		</style>
	</head>
	
	<body>
		<h1 id="heading">Welcome to the Module Admin Screen</h1>
		
		<div class = "forms" id="form1", style="display:none">
			<ul>
				<li>
					<label for="name">Module Name:</label>
					<input type="text" id="name" name="modName">
				</li>
				<li>
					<label for="code">Module Code:</label>
					<input type="text" id="code" name="modCode">
				</li>
				<li>
					<label for="admins">Admins:</label>
					<input type="text" id="admins" name="admins">
				</li>
				<li>
					<label for="TAs">Num TAs:</label>
					<input type="text" id="TAs" name="numTa">
				</li>
				<li>
					<label for="stud">Num Students:</label>
					<input type="text" id="stud" name="numStudents">
				</li>
				<li>
					<label for="reqType">Req. Types:</label>
					<input type="text" id="reqType" name="reqTypes">
				</li>
				<li>
					<label for="groups">Groups:</label>
					<input type="text" id="groups" name="groups">
				</li>
				<li>
					<label for="hours">Lab Hrs:</label>
					<input type="text" id="hours" name="hours">
				</li>
				
				<li class="button">
				<button id="button2" type="button", onclick="submit()">Submit</button>
				</li>
			</ul>
		</div>

		<button id="button1", onclick="dispForm()">Add Module</button>
		<p id="showData"></p>
		

		<script>
		console.log('outside func call');
		function createTab (tableData, flag) {

			// create the headings of the table (fixed)
			console.log('table fun!');
			console.log('DATA sent: ' + tableData.modules);

			// check if the modules list is empty 

			if (flag == false){
				console.log('clear table');
				var tab = document.getElementById('showData');
				tab.style.display = 'none';
				var warningMessage = document.createElement('header');
				warningMessage.setAttribute('id', 'noModMsg');
				text = document.createElement('h2');
				text.textContent = 'No modules to display. Please add a new module.'; 
				warningMessage.appendChild(text);
				document.body.appendChild(warningMessage);  
				var msg = document.getElementById('noModMsg');
				msg.style.display = '';	 
			}
				
			else {
			// clear warning message 
			var msg = document.getElementById('noModMsg');
			if (msg){
				msg.parentNode.removeChild(msg);
			}

			var tabHeadings = Object.keys(tableData.modules[0]);
			tabHeadings.sort(); 
			var table = document.createElement('table'); 
			var tr = table.insertRow(-1);
			for (var i in tabHeadings) {
				var th = document.createElement('th');
				th.innerHTML = tabHeadings[i].substring(1);
				tr.appendChild(th); 
			};
			

			// for each module in the JSON data create a new row 
			// add the module data as a list element
        	for (let i in tableData.modules) {
				var trInternal = table.insertRow(-1);
				trInternal.setAttribute('id', i); 
				var rowId = document.getElementById(i); 
				console.log('set row ID');
				console.log(tableData.modules[i]);
				var keys = Object.keys(tableData.modules[i]);
 
				// populate the columns with data
				for (let j=0; j<keys.length; ++j) {
					console.log(typeof keys[j]); 
					var col = trInternal.insertCell(-1);
					col.innerHTML = tableData.modules[i][keys[j]];  
					col.setAttribute('id', j); 
					var val = keys[j];
					col.setAttribute('key', val);  
					col.setAttribute('flag', true);
					col.setAttribute('onclick', 'makeEdit(this, this.parentNode.id)');
					
				} 	
				// add remove module buttons to each row in last column
				var btnCol = trInternal.insertCell(-1); 
				var btn = document.createElement('button');
				btn.setAttribute('class', 'rmBtn'); 
				btn.setAttribute('type', 'button');
				btn.setAttribute('id', i); 
				btn.setAttribute('onclick', 'rmData(id)');  
				btn.innerHTML = 'Remove';
				btnCol.appendChild(btn); 
			
			}	

			// add extra column for the remove module button 
			var divContainer = document.getElementById('showData');
			divContainer.innerHTML = ""; 
			divContainer.appendChild(table)
			divContainer.style.display = ''; 

		}}

		function rmData (id) {
			// post message to constructors.js script to call removal function
			// of specific module object in JSON database 
			console.log(id);

			// constructors.js will remove desired entry from JSON file and then redraw the table 
			vscode.postMessage({command: "remove", info: id});

			 
		}

		function makeEdit(col, rowId){
			console.log('called function!');
			// note that js elements cannot be boolean, they're all strings ... 
			if (col.getAttribute('flag') == 'true'){
				var field = document.createElement('input'); 
				field.setAttribute("type","text");
				field.setAttribute("size","10"); 
				console.log('inside func: '+ col.innerHTML);
				field.defaultValue = col.innerHTML;

				// get rid of the table text and replace with input field 
				col.innerHTML = '';
				col.appendChild(field);

				// set a flag to stop multiple input boxes being generated 
				col.setAttribute("flag", false);

				// set event listener to wait for input to finish 
				// when user hits enter key, the new values are added 
				col.addEventListener('keypress', function(e){
					if (e.key === 'Enter'){
						console.log('enter detected!');
						// cosmetic 
						col.removeChild(field); 
						col.innerHTML = field.value;
						
						// pass the updated info to edit() function in constructors 
						vscode.postMessage({command: "edit", info: [rowId, col.getAttribute('key'), field.value]});

						// allow clickable edits again 
						col.flag = true;
					}

				
				} );

			}

			 
			
		}
		</script>
		
	</body>
	</html>`
}; 

// add module to main file 
//export{ getWelcomeScreen, getModScreen }


constructors.js
const htmlStuff = require('./htmlStuff.js');
const vscode = require('vscode');
const fs = require('fs');

function Welcome (context) {
    
    // create the welcome window 
    const screen = vscode.window.createWebviewPanel('welcome', "Lab Ticketing System",
    vscode.ViewColumn.One, {enableScripts: true}, );
        
    // read in the html for the welcome window
    screen.webview.html = htmlStuff.getWelcomeScreen();

    // handle recieving messages from the webview, which creates new screen objects depending on the button press 
    screen.webview.onDidReceiveMessage(message => {switch(message.command){case 'modAdmin': var admin = new ModuleAdmin(); 
    return;}}, undefined, context.subscriptions); 

    // display TA login form when TA button clicked 
    screen.webview.onDidReceiveMessage(message => {switch(message.command){case 'ta': screen.webview.html = htmlStuff.getTaLogin(); 
    return;}}, undefined, context.subscriptions); 

// module admin object 
function ModuleAdmin () {

    // update html to display mod admin window
    screen.webview.html = htmlStuff.getModScreen();

    // module screen attributes 
    this.moduleLists = undefined; 

    //listen for new module info 
    screen.webview.onDidReceiveMessage(message => {switch(message.command){case 'newInfo': this.add(message.info); 
    return;}}, undefined, context.subscriptions);

    // listen for remove module call 
    screen.webview.onDidReceiveMessage(message => {switch(message.command){case 'remove': this.remove(message.info); 
    return;}}, undefined, context.subscriptions);

    // listen for edit module call 
    screen.webview.onDidReceiveMessage(message => {switch(message.command){case 'edit': this.edit(message.info); 
    return;}}, undefined, context.subscriptions); 

    // listen for go back navigation cloick 
    screen.webview.onDidReceiveMessage(message => {switch(message.command){case 'back': screen.webview.html = htmlStuff.getWelcomeScreen(); 
    return;}}, undefined, context.subscriptions); 

    // module screen methods to add, remove, load, save and display modules on the screen
    this.load = function loadModules () { 
        
        // check if text file with saved module list exists
        vscode.window.showInformationMessage('Loading modules . . .');

        // check if modules JSON object exists and if yes load it up
        try {
            console.log('Attempting to load modules. . .');
            var fileContent = fs.readFileSync('./modules.txt').toString();
            console.log('parsing JSON file to JS object. . .');
            var fileObj = JSON.parse(fileContent);
            console.log('Success! Modules are: ' + JSON.stringify(fileObj.modules));

            return fileObj; 
        }
        catch (e) {
            // failed to load as JSON object, so obj does not exist (first start-up)
            // create the required JSON struct required for later use 
            console.log('Failed! Creating JSON format -- first time startup --');
            var fileContent = '{"modules":[]}'; 
            fs.appendFileSync('modules.txt', fileContent); 
            console.log('Ready to add new modules!');

            // convert string to object in JSON format 
            fileContent = JSON.parse(fileContent); 
        } 

        return fileContent; 
    }; 
    
    this.add = function addModule (info) {

        console.log('Adding new module!'); 
        
        // push the parsed new content to the required position in JSON modules list 
        this.modulesList['modules'].push(info); 
        console.log('writing new file ...'); 

        // write updated modules list to file 
        fs.writeFileSync('./modules.txt', JSON.stringify(this.modulesList, null, ' ')); 

        // send message to webview to update the displayed table with new information 
        
        screen.webview.postMessage({command: 'drawTab', info: this.modulesList}); 

        console.log(this.modulesList)
    };

    this.edit = function editModule (info) {

        console.log('IDX of updates: ' + info[0] + info[1]); 
        // info is an array [rowId, key, updated value]
        // use index to select desired value to update 
        this.modulesList.modules[info[0]][info[1]] = info[2]; 
        
        // save updates 
        fs.writeFileSync('./modules.txt', JSON.stringify(this.modulesList, null, ' ')); 

        // redraw table 
        screen.webview.postMessage({command: 'drawTab', info: this.modulesList}); 
    }; 

    this.remove = function removeModule (info) {
        
        console.log('remove index: ' + info); 

        // remove desired module from the JSON file list
        this.modulesList.modules.splice(info, 1);
        console.log('length: ' + this.modulesList.modules.length);
        // update the JSON file and redraw table 
        // write updated modules list to file 
        fs.writeFileSync('./modules.txt', JSON.stringify(this.modulesList, null, ' ')); 

        // send message to webview to update the displayed table with new information 
        if (this.modulesList.modules.length < 1){
            // don't draw the table
            console.log("don't draw!");
            screen.webview.postMessage({command: 'drawTab', info: this.modulesList, draw: false});
        }
        else {
            screen.webview.postMessage({command: 'drawTab', info: this.modulesList, draw:true}); 
        }
        
        
    }; 

    // load current modules list and add contents of module file to class attributes 
    // this allows moduleList to be accessed and read from all methods 
    this.modulesList = this.load();

    // send message to webview to update the displayed table with new information 
    if (this.modulesList.modules.length < 1){
        // don't draw the table
        console.log("don't draw!");
        screen.webview.postMessage({command: 'drawTab', info: this.modulesList, draw: false});
    }
    else {
        screen.webview.postMessage({command: 'drawTab', info: this.modulesList, draw:true}); 
    }
}}

// set up exports for module 
module.exports = { Welcome }




htmlStuff.js
function getWelcomeScreen () {
	return `<!DOCTYPE html>
	<html lang="en">
	<head>
		<!-- Add icon library -->
		<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
		<script>
		const vscode = acquireVsCodeApi();
		</script>

		<meta charset="UTF-8">
		<meta name="viewport" content="width=device-width, initial-scale=1.0">
		
		<title>Module Manager</title>
		
		<style> 
		.buttons {
			position: absolute; 
			top: 40% ;
			left: 50%;
			transform: translate(-50%, 0%);
		}
		/* Style buttons */
		.btn {
		  background-color: DodgerBlue; /* Blue background */
		  border: none; /* Remove borders */
		  color: white; /* White text */
		  padding: 12px 16px; /* Some padding */
		  font-size: 16px; /* Set a font size */
		  cursor: pointer; /* Mouse pointer on hover */
		}
		
		/* Darker background on mouse-over */
		.btn:hover {
		  background-color: RoyalBlue;
		} 
		body {
			background-color: rgb(60,99,201);}

		h1 {
			position: absolute; 
			top: 30%;
			left: 50%;
			transform: translate(-50%, -50%);
			color: white;}
		
		

		

		</style>
	</head>
	
	<body>
		<div class="buttons">
		<button class="btn" onclick="moduleAdd()"><i class="fa fa-key"></i> Admins</button>
		<button class="btn" onclick="taLogin()"><i class="fa fa-user-circle"></i> TAs</button>
		<button class="btn"><i class="fa fa-graduation-cap"></i> Students</button>
		</div>

		<h1>Welcome to the Lab Ticketing System</h1>

		<script>
			function moduleAdd() {
			vscode.postMessage({command: 'modAdmin', text: "BUTTON PRESSED!"});
			}

			function taLogin() {
				console.log('ta button pressed!');
				vscode.postMessage({command:'ta'});
			}
		</script> 
		
		
		
		
		
		
	</body>
	</html>`
};

function getModScreen () {
	return `<!DOCTYPE html>
	<html lang="en">
	<head>
		<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
		<script>
		const vscode = acquireVsCodeApi();
		</script>

		<script>
			function submit () {
				var Module = {
				aModuleName     : document.getElementById('name').value,
				bModuleCode     : document.getElementById('code').value,
				cModuleAdmins   : document.getElementById('admins').value,
				dModuleTAs      : document.getElementById('TAs').value,
				eModuleStudents : document.getElementById('stud').value,
				fModuleReqType  : document.getElementById('reqType').value,
				gModuleGroups   : document.getElementById('groups').value,
				hModuleHrs      : document.getElementById('hours').value};


				vscode.postMessage({command: "newInfo", info: Module});

				//hide the module input screen, re-display add module button and current modules added  
				document.getElementById('form1').style.display = 'none';
				document.getElementById('button1').style.display = ''; 
				document.getElementById('showData').style.display ='';
				document.getElementById('back').style.display ='';
			}
		</script>

		<script>
			function dispForm () {
				document.getElementById('form1').style.display = '';
				document.getElementById('button1').style.display = 'none'; 
				// hide the module table
				document.getElementById('showData').style.display = 'none'; 
				document.getElementById('back').style.display = 'none'; 
			};
		</script>

		<script>
			// create event listener for table data fetch 
			window.addEventListener('message', event => {
				var message = event.data; 
				
				switch (message.command) {
					
					case 'drawTab':
					// call drawing function and pass the JSON data  
					console.log('Message recieved: ' +  JSON.stringify(message.info)); 
					createTab(message.info, message.draw); 
				}
			});
		</script>

		<meta charset="UTF-8">
		<meta name="viewport" content="width=device-width, initial-scale=1.0">
		
		<title>Module Manager</title>
		
		<style>
		.buttons {
			position: absolute; 
			top: 40% ;
			left: 50%;
			transform: translate(-50%, 0%);
		}
		/* Style buttons */
		.btn {
		  background-color: DodgerBlue; /* Blue background */
		  border: none; /* Remove borders */
		  color: white; /* White text */
		  padding: 12px 16px; /* Some padding */
		  font-size: 16px; /* Set a font size */
		  cursor: pointer; /* Mouse pointer on hover */
		}
		
		/* Darker background on mouse-over */
		.btn:hover {
		  background-color: RoyalBlue;
		} 
		body {
			background-color: rgb(60,99,201);}
		.forms {
			/* Center the form on the page */
			
			width: 400px;
			position: absolute; 
			top: 400px;
			right: 740px;
			/* Form outline */
			padding: 0.5em;
			border: 1px solid #CCC;
			border-radius: 1em;
		}
		  
		ul {
			list-style: none;
			padding: 0;
			margin: 0;
		}
		  
		.forms li + li {
			margin-top: 1em;
		}
		  
		label {
			/* Uniform size & alignment */
			display: inline-block;
			width: 90px;
			text-align: right;
		} 
		body {
			background-color: rgb(60,99,201);}

		h1 {
			text-align: center;
			position: absolute; 
			top: 30%;
			left: 50%;
			transform: translate(-50%, -50%);
			color: white;}

		#noModMsg {
			text-align: center; 
			position: absolute; 
			top: 40%; 
			left: 50%;
			transform: translate(-50%, -50%);
		}

		
		#button2 {
			position: relative;
			left: 52%;
		}

		table {
			position: absolute; 
			top: 50%; 
			left: 31%;
			border-collapse: collapse;}
		tr { 
			border: solid;
			border-width: 1px 0;
			}
		th, td{ 
			border-left:1px solid #ccc; 
			border-right: 1px solid #ccc;
		}

		.rmBtn {
			background-color: rgb(51, 255, 255);
			color: rgb(51, 54, 255);
			border: border-width border-style border-color|initial|inherit;
			border-color: rgb(51, 255, 255);
		}
		</style>
	</head>
	
	<body>
		<h1 id="heading">Welcome to the Module Admin Screen</h1>
		
		<div class = "forms" id="form1", style="display:none">
			<ul>
				<li>
					<label for="name">Module Name:</label>
					<input type="text" id="name" name="modName">
				</li>
				<li>
					<label for="code">Module Code:</label>
					<input type="text" id="code" name="modCode">
				</li>
				<li>
					<label for="admins">Admins:</label>
					<input type="text" id="admins" name="admins">
				</li>
				<li>
					<label for="TAs">Num TAs:</label>
					<input type="text" id="TAs" name="numTa">
				</li>
				<li>
					<label for="stud">Num Students:</label>
					<input type="text" id="stud" name="numStudents">
				</li>
				<li>
					<label for="reqType">Req. Types:</label>
					<input type="text" id="reqType" name="reqTypes">
				</li>
				<li>
					<label for="groups">Groups:</label>
					<input type="text" id="groups" name="groups">
				</li>
				<li>
					<label for="hours">Lab Hrs:</label>
					<input type="text" id="hours" name="hours">
				</li>
				
				<li class="button">
				<button id="button2" type="button", onclick="submit()">Submit</button>
				</li>
			</ul>
		</div>

		<div class="buttons">
		<button class="btn" id="back" onclick="back()"><i class="fa fa-backward "></i> Back</button>
		<button class="btn" id="button1" onclick="dispForm()"><i class="fa fa-plus"></i> Add Module</button>
		</div>
		<p id="showData"></p>
		<script>
			function back(){
				vscode.postMessage({command: 'back'}); 
			} 
		</script>
		<script>
		console.log('outside func call');
		function createTab (tableData, flag) {

			// create the headings of the table (fixed)
			console.log('table fun!');
			console.log('DATA sent: ' + tableData.modules);

			// check if the modules list is empty 

			if (flag == false){
				console.log('clear table');
				var tab = document.getElementById('showData');
				tab.style.display = 'none';
				var warningMessage = document.createElement('header');
				warningMessage.setAttribute('id', 'noModMsg');
				text = document.createElement('h2');
				text.textContent = 'No modules to display. Please add a new module.'; 
				warningMessage.appendChild(text);
				document.body.appendChild(warningMessage);  
				var msg = document.getElementById('noModMsg');
				msg.style.display = '';	 
			}
				
			else {
			// clear warning message 
			var msg = document.getElementById('noModMsg');
			if (msg){
				msg.parentNode.removeChild(msg);
			}

			var tabHeadings = Object.keys(tableData.modules[0]);
			tabHeadings.sort(); 
			var table = document.createElement('table'); 
			var tr = table.insertRow(-1);
			for (var i in tabHeadings) {
				var th = document.createElement('th');
				th.innerHTML = tabHeadings[i].substring(1);
				tr.appendChild(th); 
			};
			

			// for each module in the JSON data create a new row 
			// add the module data as a list element
        	for (let i in tableData.modules) {
				var trInternal = table.insertRow(-1);
				trInternal.setAttribute('id', i); 
				var rowId = document.getElementById(i); 
				console.log('set row ID');
				console.log(tableData.modules[i]);
				var keys = Object.keys(tableData.modules[i]);
 
				// populate the columns with data
				for (let j=0; j<keys.length; ++j) {
					console.log(typeof keys[j]); 
					var col = trInternal.insertCell(-1);
					col.innerHTML = tableData.modules[i][keys[j]];  
					col.setAttribute('id', j); 
					var val = keys[j];
					col.setAttribute('key', val);  
					col.setAttribute('flag', true);
					col.setAttribute('onclick', 'makeEdit(this, this.parentNode.id)');
					
				} 	
				// add remove module buttons to each row in last column
				var btnCol = trInternal.insertCell(-1); 
				var btn = document.createElement('button');
				btn.setAttribute('class', 'btn'); 
				btn.setAttribute('type', 'button');
				btn.setAttribute('id', i); 
				btn.setAttribute('onclick', 'rmData(id)');  
				btn.innerHTML = '<i class="fa fa-minus"> Remove</i>';
				btnCol.appendChild(btn); 
			
			}	

			// add extra column for the remove module button 
			var divContainer = document.getElementById('showData');
			divContainer.innerHTML = ""; 
			divContainer.appendChild(table)
			divContainer.style.display = ''; 

		}}

		function rmData (id) {
			// post message to constructors.js script to call removal function
			// of specific module object in JSON database 
			console.log(id);

			// constructors.js will remove desired entry from JSON file and then redraw the table 
			vscode.postMessage({command: "remove", info: id});

			 
		}

		function makeEdit(col, rowId){
			console.log('called function!');
			// note that js elements cannot be boolean, they're all strings ... 
			if (col.getAttribute('flag') == 'true'){
				var field = document.createElement('input'); 
				field.setAttribute("type","text");
				field.setAttribute("size","10"); 
				console.log('inside func: '+ col.innerHTML);
				field.defaultValue = col.innerHTML;

				// get rid of the table text and replace with input field 
				col.innerHTML = '';
				col.appendChild(field);

				// set a flag to stop multiple input boxes being generated 
				col.setAttribute("flag", false);

				// set event listener to wait for input to finish 
				// when user hits enter key, the new values are added 
				col.addEventListener('keypress', function(e){
					if (e.key === 'Enter'){
						console.log('enter detected!');
						// cosmetic 
						col.removeChild(field); 
						col.innerHTML = field.value;
						
						// pass the updated info to edit() function in constructors 
						vscode.postMessage({command: "edit", info: [rowId, col.getAttribute('key'), field.value]});

						// allow clickable edits again 
						col.flag = true;
					}

				
				} );

			}

			 
			
		}
		</script>
		
	</body>
	</html>`
}; 

function getTaLogin() {
	return `<!DOCTYPE html> 
	<head>
	<script>
		const vscode = acquireVsCodeApi();
	</script>
		<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
	<style>
		body {
			background-color: rgb(60,99,201);}
		
		.login {
			display: none; /* Hidden by default */
  			position: fixed; /* Stay in place */
			z-index: 1; /* Sit on top */
			left: 0;
			top: 0;
			width: 100%; /* Full width */
			height: 100%; /* Full height */
			overflow: auto; /* Enable scroll if needed */
			background-color: rgb(0,0,0); /* Fallback color */
			background-color: rgba(0,0,0,0.4); /* Black w/ opacity */
			padding-top: 60px;
		}
	</style>
	</head>

	<body>
	
		<h1>TA Login Screen!</h1>
	
	</body>
	
	</html>`
}

// set up exports for module
module.exports = { getWelcomeScreen, getModScreen, getTaLogin }
